### PR TITLE
[SE-0280] Enum cases as protocol witnesses

### PR DESCRIPTION
**Pull request by theblixguy**
Allow enum cases to witness static protocol requirements. For example:

```swift
protocol Foo {
  static var bar: Self { get }
  static func baz(arg: Int) -> Self
}

enum SomeEnum: Foo {
  case bar // Okay
  case baz(arg: Int) // Okay
}
```

Swift-evolution thread: https://forums.swift.org/t/pitch-enum-cases-as-protocol-witnesses/32753

Resolves SR-3170, SR-3510, SR-9099

### CODE REVIEW

**Comment by theblixguy** [comment]
ping @slavapestov

File: lib/Sema/TypeCheckProtocol.cpp
Comment (slavapestov) [inline lib/Sema/TypeCheckProtocol.cpp]: I think var->isMutable() would be a better query for this.

**Review by slavapestov — APPROVED** [review]
I'm OK with this as long as the core team feels it doesn't need their approval.

CC @DougGregor @rjmccall @jckarter

**Comment by rjmccall** [comment]
Hmm.  We might.  I'll raise it to the Core Team.

**Comment by theblixguy** [comment]
@rjmccall Any update on this from the core team?

File: lib/Sema/TypeCheckProtocol.cpp
Comment (theblixguy) [reply lib/Sema/TypeCheckProtocol.cpp]: There is no `isMutable()` method on `VarDecl` unfortunately.

**Comment by rjmccall** [comment]
No, sorry.  It slipped through the cracks during the last meeting, but I made sure it went back on the agenda for tomorrow.

**Comment by theblixguy** [comment]
Thank you!

**Comment by theblixguy** [comment]
@rjmccall Okay, sounds good! I did manage to get it to work with cases with payloads (I'll update this PR) while I was waiting for a response so I think the proposal + implementation part is covered and I can pitch it on the forums.

In terms of the "mini-manifesto", I am happy to write one down based on the points you've provided and pitch on the forums. Does this manifesto need to be "approved" by the core team before the proposal can be considered for review?

**Comment by rjmccall** [comment]
> @rjmccall Okay, sounds good! I did manage to get it to work with cases with payloads (I'll update this PR) while I was waiting for a response so I think the proposal + implementation part is covered and I can pitch it on the forums.

That's awesome and great to hear.

> In terms of the "mini-manifesto", I am happy to write one down based on the points you've provided and pitch on the forums. Does this "mini-manifesto" need to be "approved" by the core team before the proposal can be considered for review?

We'll need to talk about what process we want to use for this, but you can go ahead and start the conversation in the meantime, I think.

Thanks for taking this on!

**Comment by theblixguy** [comment]
Hmm, I wanted to build a toolchain, but it seems like there's some sort of linking issue because when the witness is a payload case and (1) I build with `-Onone` and/or (2) I use `public` visibility then I get:

```
Undefined symbols for architecture x86_64:
  "_$s4test1EO3baryACSi_tcACmF", referenced from: // enum case constructor
      _$s4test1EOAA1PA2aDP3bar5valuexSi_tFZTW in test-2582e0.o // witness thunk
ld: symbol(s) not found for architecture x86_64
```

I don't get this with `-O` unless I use `public`. It doesn't affect payload-less case witness, only ones with payload. It's a little strange.... there's probably a few places where we're still assuming the witness could only be a func/var/constructor I think.

**Comment by jckarter** [comment]
@theblixguy I believe we only generate the case constructor symbols by need as shared-linkage symbols so that they don't have to be ABI. You might need to trigger the case constructor's generation when it's used as a witness.

**Comment by theblixguy** [comment]
@jckarter Thank you for the tip! I inserted a call to `emitEnumConstructor` before the call to `emitProtocolWitness`, however in this case, `SILDeclRef::getLinkage` never actually returns shared linkage. It falls down and returns hidden_external linkage. I tweaked it to handle an enum element decl and set the `limit` to `OnDemand` which fixes the problem. I am not sure if that's the right/correct way though. Without this, I kept running into issues where `emitEnumConstructor` will emit it with shared linkage, but `emitProtocolWitness` would use hidden_external linkage.

**Comment by slavapestov** [comment]
Enum case constructors should be emitted as needed already by getGlobalFunctionRef(). You can reference a case constructor across modules already, via currying syntax. Can you see why it's not working when referenced from the witness thunk code path?

**Comment by slavapestov** [comment]
Eg, if I write

```
_ = Optional<Int>.some
```

It emits two thunks:

```
sil shared [transparent] [serializable] [ossa] @$sSq4someyxSgxcABmlF : $@convention(method) <Wrapped> (@in Wrapped, @thin Optional<Wrapped>.Type) -> @out Optional<Wrapped> {
```

This is the thunk emitted by emitEnumConstructor(). It has shared linkage and a body.

Then there's the curry thunk which gives it the AST type `(Optional<Int>.Type) -> (Int) -> (Optional<Int>.some)`:

```
sil shared [transparent] [serializable] [thunk] [ossa] @$sSq4someyxSgxcABmlFTc : $@convention(thin) <Wrapped> (@thin Optional<Wrapped>.Type) -> @owned @callee_guaranteed (@in_guaranteed Wrapped) -> @out Optional<Wrapped> {
```

**Comment by theblixguy** [comment]
@slavapestov Yeah, I think I have now found the actual problem! So, it's my fault, I made a mistake in `SILGenWitnessTable::addMethod`. It seems like the code I added to call `addMethodImplementation` wasn't being invoked in some cases, so we ended up with a `SILDeclRef` to an enum case witness, but a `Func` kind. This also explains why I had to manually return a function type in `getFunctionInterfaceTypeWithCaptures`!

**Comment by theblixguy** [comment]
@swift-ci please build toolchain

**Comment by theblixguy** [comment]
@swift-ci please build toolchain

File: lib/TBDGen/TBDGen.cpp
Comment (slavapestov) [inline lib/TBDGen/TBDGen.cpp]: In practice, enum constructors never have public linkage so they don't need to be visited by TBDGen. However in case this ever changes, there's one less spot to update...

**Review by slavapestov — APPROVED** [review]
Glad you sorted out the problem with emitting enum constructors. The change looks good to me now.

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by theblixguy** [comment]
@swift-ci please test

**Comment by theblixguy** [comment]
@swift-ci please test

**Comment by theblixguy** [comment]
@swift-ci please test

**Comment by theblixguy** [comment]
@swift-ci please test

**Comment by theblixguy** [comment]
@swift-ci please test

**Comment by theblixguy** [comment]
@swift-ci please test source compatibility

**Comment by damienissa** [comment]
Hi all. I have found some issues!

```Swift
import Foundation

public protocol Foo {
    
    static var foo: Self { get }
}

enum Bar: Foo {
    
    case foo
}

func load<F: Foo>(_ foo: F) {
    
    print(foo)
}

load(Bar.foo)

enum Bar2 {
    case foo
    case bar(Int)
}

func load2(_ bar2: Bar2) {
    
    switch bar2 {
    case .foo: break
    case let .bar(val): print(val)
    }
}

load2(.bar(5))

protocol Foo3: Equatable {
    
    static var foo: Self { get }
    static func bar(_ value: Int) -> Self
}

enum Bar3: Foo3 {
    
    case foo
    case bar(Int)
}

func load3<F3: Foo3>(_ foo3: F3) {
    
    switch foo3 {
    case .foo: break
    case let .bar(val): print(val); #error("Pattern variable binding cannot appear in an expression")
    default: break
    }
}

load3(Bar3.bar(5))
```

**Comment by xwu** [comment]
Hi @damienissa,

That behavior is correct and also unrelated to this proposal feature. (Incidentally, even if it were both incorrect and related to this proposal, the forums and/or bug tracker would be the place to discuss it.)

It is unrelated to this feature because this is about enum cases serving as witnesses for protocol requirements; this does not mean that static protocol requirements are equivalent to enum cases. It is correct behavior because the protocol does not guarantee any way of binding the argument passed to `bar`: the only thing we know about types that conform to the protocol is that we can invoke `bar` with an argument and get back a result of type `Self`. We can test to see if a given instance is equivalent to `.bar(5)` but the only way to know what integer was used to create that instance would be to test every value from `Int.min` to `Int.max`—and that’s if `bar` is a pure function.

**Comment by damienissa** [comment]
@xwu Thanks!

**Pull request merged**
